pkg/controller/deployment/deployment_controller.go

```diff
@@ -419,17 +430,52 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 	return fmt.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
 
-func (dc *DeploymentController) syncRecreateDeployment(deployment extensions.Deployment) error {
-	newRC, err := dc.getNewRC(deployment)
+// Rolling back to a revision; no-op if the toRevision is deployment's current revision
+func (dc *DeploymentController) rollback(deployment extensions.Deployment, toRevision int) error {
+	if toRevision <= 0 {
+		return nil
+	}
+	newRC, oldRCs, err := dc.getNewRCAndOldRCs(deployment, false)
 	if err != nil {
 		return err
 	}
+	allRCs := append(oldRCs, newRC)
+	for _, rc := range allRCs {
+		v, err := revision(rc)
+		if err != nil {
+			return fmt.Errorf("unable to extract revision from deployment's RCs: %v", err)
+		}
+		glog.V(4).Infof("Found rc with revision %d: %+v", v, rc)
+		if v == toRevision {
+			glog.V(4).Infof("Found a match!")
+			// rollback by copying podTemplate.Spec from the rc, and increment revision number by 1
+			// no-op if the the spec matches current deployment's podTemplate.Spec
+			if !api.Semantic.DeepEqual(deployment.Spec.Template.Spec, rc.Spec.Template.Spec) {
+				glog.Infof("Rolling back deployment %v to revision %d", deployment.Name, toRevision)
+				deployment.Spec.Template.Spec = rc.Spec.Template.Spec
+				maxV, err := maxRevision(allRCs)
+				if err != nil {
+					return fmt.Errorf("unable to retrieve the max revision of all RCs: %v", err)
```

Comment:
Same comment as on the other PR: We should tolerate missing/bad revision annotations, provided that we found the matching one.